### ISSUE TITLE
Number of given arguments should be checked in the method wrapper function

### ISSUE THREAD

**Issue by toasteater** [GodotNativeTools/godot-cpp#367]
Here in `__wrapped_method`, it should be checked that `num_args` is equal to, or at least equal to or larger than `sizeof...(As)`: https://github.com/GodotNativeTools/godot-cpp/blob/master/include/core/Godot.hpp#L197

If `num_args` is less than the number of arguments expected by the wrapped method, invalid reads will happen, because in `_WrappedMethod::apply`, the arguments are dereferenced without checking the bounds:

```cpp
template <int... I>
void apply(Variant *ret, T *obj, Variant **args, __Sequence<I...>) {
	*ret = (obj->*f)(_ArgCast<As>::_arg_cast(*args[I])...);
}
```

**Comment by Zylann** [comment]
I wonder if it should be checked by the C++ bindings. There seems to be an issue already with telling Godot which arguments the function has in the first place, Godot does not autocompletes them either https://github.com/GodotNativeTools/godot-cpp/issues/288

If we could set argument name, type and count at registration time, it would then allow the engine to do that check in the script editor and at runtime for GDScript (and have better context to report it), and also allow to generate complete bindings if it were to be used in C#... which means there won't be any need for all the GDNative libraries to add that check.

**Comment by toasteater** [comment]
I agree that argument info should be given to the engine, but I'm not sure if the [current shape of the `set_method_argument_information` function](https://github.com/godotengine/godot/blob/master/modules/gdnative/nativescript/godot_nativescript.cpp#L228) is very usable for the purpose. Most glaringly, it provides no way to set `PROPERTY_USAGE_NIL_IS_VARIANT` for the arguments, which means that if the resulting `MethodInfo` is enforced, there will be no way to explicitly accept a `Variant` from the engine. There are also no ways to add optional arguments, that are possible today with argument count handled on the library side.

Even if a new API that addresses these problems is released in 4.0 (which will take a long time), I'd still argue that it's best to add a check on the library side, because it's cheap protection from obvious UB.

Also, if I recall correctly, the information set by the `_documentation` functions isn't currently used by the engine. A search for `get_method_documentation`, for example, yields only results in the GDNative module: https://github.com/godotengine/godot/search?q=get_method_documentation&unscoped_q=get_method_documentation

**Comment by Zylann** [comment]
Turns out that was part of a very old PR I did in the era where only `GodotScript` existed: https://github.com/GodotNativeTools/godot-cpp/pull/92#issue-167816988
I may have to do some speleology to remember how that worked
